receiver/nats.go

```diff
@@ -0,0 +1,163 @@
+/*
+ * skogul, nats receiver/subscriber
+ *
+ * Author(s):
+ *  - Niklas Holmstedt <n.holmstedt@gmail.com>
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ * 02110-1301  USA
+ */
+
+package receiver
+
+import (
+	"crypto/tls"
+	"fmt"
+	"github.com/nats-io/nats.go"
+	"github.com/telenornms/skogul"
+	"sync"
+)
+
+var natsLog = skogul.Logger("receiver", "nats")
+
+/*
+Nats.io simple pub/sub receiver with:
+
+- Authentication: Username/Password, TLS
+- Authorization: Username/Password, UserCredentials/JWT
+- Queue: Load balancing for multiple receivers in the same Queue.
+*/
+type Nats struct {
+	Handler       skogul.HandlerRef `doc:"Handler used to parse, transform and send data."`
+	Servers       string            `doc:"Comma separated list of nats URLs"`
+	Subject       string            `doc:"Subject to subscribe to messages on"`
+	Queue         string            `doc:"Worker queue to distribute messages on"`
+	Name          string            `doc:"Client name"`
+	Username      string            `doc:"Client username"`
+	Password      string            `doc:"Client password"`
+	TLSClientKey  string            `doc:"TLS client key file path"`
+	TLSClientCert string            `doc:"TLS client cert file path"`
+	TLSCACert     string            `doc:"CA cert file path"`
+	UserCreds     string            `doc:"Nats credentials file path"`
+	NKeyFile      string            `doc:"Nats nkey file path"`
+	Insecure      bool              `doc:"TLS InsecureSkipVerify"`
+	o             *[]nats.Option
+	nc            *nats.Conn
+	wg            sync.WaitGroup
+}
+
+// Verify configuration
+func (n *Nats) Verify() error {
+	if n.Handler.Name == "" {
+		return skogul.MissingArgument("Handler")
+	}
+	if n.Subject == "" {
+		return skogul.MissingArgument("Subject")
+	}
+	//User Credentials
+	if n.UserCreds != "" && n.NKeyFile != "" {
+		//Cred file contains nkey.
+		return fmt.Errorf("Please configure usercreds or nkeyfile.")
+	}
+
+	return nil
+}
+
+func (n *Nats) Start() error {
+	if n.Name == "" {
+		n.Name = "skogul"
+	}
+	n.o = &[]nats.Option{nats.Name(n.Name)}
+
+	if n.Servers == "" {
+		n.Servers = nats.DefaultURL
+	}
```

Comment:
I'm not sure what the default URL of nats is, and if it is common to omit the URL used for connecting.

In general I'm more for being explicit, so I'd suggest making "Servers" a required configuration option (by adding a `Verify()` check which fails if `n.Servers` is not configured.)

Reply:
Same as above. It's pretty much how they do it everywere. I don't really have any issue with changing it tho.

It defaults to localhost.